### PR TITLE
[TT-6575] feat: persisted queries

### PR DESCRIPTION
**Pull request by kofoworola**
## Description
This PR adds a new middleware `PersistGraphQLOperationMiddleware` by @asoorm. This middleware allows the exposing graphql endpoints as Rest endpoints vie extended paths. It also allows using variables from the context variables, path or other tyk template variables. 
[Link to demo video](https://www.loom.com/share/684e1e9902054d978897e24d7824dec5)
Sample Spec:
```.json
"extended_paths": {
          "persist_graphql": [
            {
              "method": "GET",
              "path": "/getContinentByCode",
              "operation": "query ($continentCode: ID!) {\n    continent(code: $continentCode) {\n        code\n        name\n        countries {\n            name\n        }\n    }\n}",
              "variables": {
                "continentCode": "EU"
              }
            },
            {
              "method": "GET",
              "path": "/getCountryByCode",
              "operation": "query ($countryCode: ID!) {\n    country(code: $countryCode) {\n        code\n        name\n    }\n}",
              "variables": {
                "countryCode": "$tyk_context.headers_Code"
              }
            }
          ]
```

## Related Issue

[TT-6575](https://tyktech.atlassian.net/browse/TT-6575)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit:  
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3310399541)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: 33b60b8d158b76eb470089057dca342427a15c43 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3310412363)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: 3b442695a5eafc1b5ac54aae66723e913c3b0790 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3310579515)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: 518ecaaec33bbfa3c971a8ef0def0050805625b5 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3310799026)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: f7c1f62afcec81c32a57691137a5f9dfdb327cc8 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3311155765)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: 7c75848c759691e4c2a92644026a7301e39f6020 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3319518632)

**Review by pvormste — APPROVED** [review]
LGTM!

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: f52628ef57b97a4d7bab3200a9658c510acf534c 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3327508033)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: b12838d4edd5cb09a6324e663e8fee298b13e3bb 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3333880505)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: 597489561cd9b34f182206c4acb1b495efd4a865 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3333894147)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: fc69c947d9b1bd8f888f1caafd204e775aa771d2 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3335225291)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: 94485cd2789e292e25c0d52a0dd383a25094b2bd 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3335959461)

File: gateway/mw_persist_graphql_operation.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+type PersistGraphQLOperationMiddleware struct {
+	BaseMiddleware
+}
+
+func (i *PersistGraphQLOperationMiddleware) Name() string {
+	return "PersistGraphQLOperationMiddleware"
+}
+
+func (i *PersistGraphQLOperationMiddleware) EnabledForSpec() bool {
+	return true
```
Comment (buraksezer) [inline gateway/mw_persist_graphql_operation.go:24]: This always returns true which means the persistent query feature is always enabled, right? Do we need to add a new config parameter to the API def?

Comment (kofoworola) [reply gateway/mw_persist_graphql_operation.go:24]: Oh no it's always enabled, the version and extended path are what determine if it's going to be used on a request.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: 61e1110722489106e418f939612c274870d22151 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3338065689)

**Review by buraksezer — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4359/merge
Commit: df429e4a920fc28f94be34544124e16632ea0665 
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3344137361)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: eac6e1ac3625257bf8ccd6209d500b6d60a078eb [TT-6575] feat: persisted queries (#4359)

* setting up the base middleware

* dynamically injecting variables

* added tests for default version and tests for context variables

* added test for version

* added path variable test, fixed bug with variable replace

* modified tests

* moved location of middleware

* handled unhandled errors.

Co-authored-by: asoorm <ahmet@mangomm.co.uk>
Triggered by: push (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/3344391982)

**Comment by asoorm** [comment]
closes #4298

**Pull request merged**